packages/components/src/panel/body.js

```diff
@@ -2,94 +2,132 @@
  * External dependencies
  */
 import classnames from 'classnames';
+import { noop } from 'lodash';
+import mergeRefs from 'react-merge-refs';
+import {
+	useDisclosureState,
+	Disclosure,
+	DisclosureContent,
+} from 'reakit/Disclosure';
 
 /**
  * WordPress dependencies
  */
-import { Component, forwardRef } from '@wordpress/element';
+import { useReducedMotion } from '@wordpress/compose';
+import { forwardRef, useRef } from '@wordpress/element';
 import { chevronUp, chevronDown } from '@wordpress/icons';
 
 /**
  * Internal dependencies
  */
 import Button from '../button';
 import Icon from '../icon';
+import { useUpdateEffect } from '../utils';
 
-export class PanelBody extends Component {
-	constructor( props ) {
-		super( ...arguments );
-		this.state = {
-			opened: props.initialOpen === undefined ? true : props.initialOpen,
-		};
-		this.toggle = this.toggle.bind( this );
-	}
+export function PanelBody(
+	{
+		children,
+		className,
+		focusable,
+		icon,
+		initialOpen: initialOpenProp,
+		onToggle = noop,
+		opened,
+		title,
+	},
+	ref
+) {
+	const initialOpen = useRef( initialOpenProp ).current;
+	const disclosure = useDisclosureState( {
+		visible: initialOpen !== undefined ? initialOpen : opened,
+	} );
+	const nodeRef = useRef();
 
-	toggle( event ) {
-		event.preventDefault();
-		if ( this.props.opened === undefined ) {
-			this.setState( ( state ) => ( {
-				opened: ! state.opened,
-			} ) );
-		}
+	// Defaults to 'smooth' scrolling
+	// https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView
+	const scrollBehavior = useReducedMotion() ? 'auto' : 'smooth';
+	const isOpened = disclosure.visible;
 
-		if ( this.props.onToggle ) {
-			this.props.onToggle();
+	// Runs after initial render
+	useUpdateEffect( () => {
+		if ( disclosure.visible ) {
+			/*
+			 * Scrolls the content into view when visible.
+			 * This improves the UX when there are multiple stacking <PanelBody />
+			 * components in a scrollable container.
+			 */
+			if ( nodeRef.current.scrollIntoView ) {
+				nodeRef.current.scrollIntoView( {
+					inline: 'nearest',
+					block: 'nearest',
+					behavior: scrollBehavior,
+				} );
+			}
 		}
-	}
 
-	render() {
-		const {
-			title,
-			children,
-			opened,
-			className,
-			icon,
-			forwardedRef,
-		} = this.props;
-		const isOpened = opened === undefined ? this.state.opened : opened;
-		const classes = classnames( 'components-panel__body', className, {
-			'is-opened': isOpened,
-		} );
+		onToggle( disclosure.visible );
+	}, [ disclosure.visible, onToggle, scrollBehavior ] );
```

Comment:
The `onToggle` here on the deps array means that this prop should always be memoized. For example, this would make this effect run on every render:
```jsx
<PanelBody onToggle={() => {}} />
```
The user would have to do this:
```jsx
const onToggle = useCallback(() => {}, []);
<PanelBody onToggle={onToggle} />
```
Is this intended?

Reply:
@diegohaz Thanks for pointing this out! I'll remove it from the deps array

Reply:
I think removing from the deps array is not enough as you could end up with a component calling an old version of the "onToggle" prop. I solve this using refs personally. Example https://github.com/WordPress/gutenberg/pull/23394/files#diff-9db94cc958a10efcb72038925b6bc24cR57-R60

Reply:
Hmm! Interesting :). Thanks for the suggestion! If this pattern continues to pop up, maybe we can create a util for this effect :D

Reply:
I use [this helper](https://github.com/reakit/reakit/blob/9bd6882fb50948c616e757d750b436decc36ddfc/packages/reakit-utils/src/useLiveRef.ts) on Reakit. The difference is that it's `useLayoutEffect` so it runs before all `useEffect`s and also other `useLayoutEffect`s if it's called first. And there's no dependency array.